pkg/openshift/osd/upgrade.go

```diff
@@ -0,0 +1,351 @@
+func (o *Provider) getVersionGateAgreement(ctx context.Context, versionGateID string) (*clustersmgmtv1.VersionGate, error) {
+	response, err := o.ClustersMgmt().V1().VersionGates().VersionGate(versionGateID).Get().SendContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed  to get version gate agreement %q, %v", versionGateID, err)
+	}
+
+	return response.Body(), nil
+}
+
+// gateAgreementExistForCluster checks to see if the version gate agreement id provided for the cluster already exists
+func (o *Provider) gateAgreementExistForCluster(ctx context.Context, clusterID, gateAgreementID string) (bool, error) {
+	response, err := o.ClustersMgmt().V1().Clusters().Cluster(clusterID).GateAgreements().List().SendContext(ctx)
+	if err != nil {
+		return false, fmt.Errorf("failed to get cluster %q version gate agreement: %v", clusterID, err)
+	}
+
+	for _, gateAgreement := range response.Items().Slice() {
+		if gateAgreement.VersionGate().ID() == gateAgreementID {
+			o.log.Info("Gate agreement exists", clusterIDLoggerKey, clusterID, "gateAgreementID", gateAgreementID)
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// addGateAgreement adds a version gate agreement to the cluster ocm resource.
+// Version gate agreement are used to acknowledge the cluster can be upgraded between versions
+func (o *Provider) addGateAgreement(ctx context.Context, clusterID string, currentVersion, upgradeVersion semver.Version) error {
+	if !(currentVersion.Minor() < upgradeVersion.Minor()) {
+		o.log.Info("Gate agreement not required for z stream upgrades", clusterIDLoggerKey, clusterID)
+		return nil
+	}
+
+	majorMinor := fmt.Sprintf("%d.%d", upgradeVersion.Major(), upgradeVersion.Minor())
+
+	versionGateID, err := o.getVersionGateID(ctx, majorMinor)
+	if err != nil {
+		return err
+	}
+
+	exist, err := o.gateAgreementExistForCluster(ctx, clusterID, versionGateID)
+	if err != nil {
+		return err
+	}
+	if exist {
+		return nil
+	}
+
+	gateAgreement, err := o.getVersionGateAgreement(ctx, versionGateID)
+	if err != nil {
+		return err
+	}
+
+	versionGateAgreement, err := clustersmgmtv1.NewVersionGateAgreement().
+		VersionGate(clustersmgmtv1.NewVersionGate().Copy(gateAgreement)).
+		Build()
+	if err != nil {
+		return fmt.Errorf("failed to build version gate agreement for cluster %q, %v", clusterID, err)
+	}
+
+	_, err = o.ClustersMgmt().V1().Clusters().Cluster(clusterID).GateAgreements().Add().Body(versionGateAgreement).SendContext(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to apply version gate agreement to cluster %q, %v", clusterID, err)
+	}
+
+	return nil
+}
+
+// initiateUpgrade initiates the upgrade for the cluster with ocm by applying a upgrade policy to the cluster
+func (o *Provider) initiateUpgrade(ctx context.Context, clusterID, version string) error {
+	upgradePolicy, err := clustersmgmtv1.NewUpgradePolicy().Version(version).
+		NextRun(time.Now().UTC().Add(7 * time.Minute)).
```

Comment:
Is there a reason this is 7 minutes and not lower?

Reply:
I tried doing < 7 minutes but the API returned an error saying it needed to be scheduled out longer. 7 minutes was pulled from osde2e. For now, I will leave it as is but look to revise it in the future if possible.